### PR TITLE
New callback interface for training visualization.

### PR DESCRIPTION
**Pull request by kevinthesun**
Neural style transfer notebook example of (refactored) live notebook charting.

Trivial notebook to run train_cifar10 working.

notebook structure in place for live charting.
Needs module cb change.

Module callbacks almost working.  Epoch numbers NaN

Module / CNN example for notebook charts.

CIFAR10 training notebook for module works again.

Matching callback code to notebook.

Add new notebook

Small fix

Change notebook name

Clean merge

### CODE REVIEW

File: python/mxnet/model.py [hunk 1]
```diff
@@ -112,13 +112,40 @@ def _update_params(param_arrays, grad_arrays, updater, num_device,
...
+    """
+    if isinstance(callbacks, list):
+        for cb in callbacks:
+            cb(*args, **kwargs)
+        return
+    if callbacks:
+        callbacks(*args, **kwargs)
+
+
+class LogValidationMetricsCallback(object):
```
Comment (piiswrong) [inline python/mxnet/model.py]: move to mx.callback

File: python/mxnet/notebook/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+import bokeh
```
Comment (piiswrong) [inline python/mxnet/notebook/__init__.py]: this doesn't come with python by default. Should be imported optionally

File: python/mxnet/notebook/callback.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
+import datetime
+import bokeh.plotting
+from collections import defaultdict
+import pandas as pd
```
Comment (piiswrong) [inline python/mxnet/notebook/callback.py]: all these should be imported optionally

File: python/mxnet/model.py (hunk 1 above)
Comment (kevinthesun) [inline python/mxnet/model.py]: Move LogValidationMetricsCallback class to callback.py?

File: python/mxnet/notebook/__init__.py (hunk 2 above)
Comment (leopd) [inline python/mxnet/notebook/__init__.py]: For optional import, do you mean like how opencv is loaded here?https://github.com/dmlc/mxnet/blob/master/python/mxnet/recordio.py#L18

**Comment by piiswrong** [comment]
Yeah
```
try:
    import bokeh
except ImportError:
    bokeh = None
```
Is cleaner. I changed it for cv2 in nnvm branch

**Comment by kevinthesun** [comment]
Also what do you mean for moving to mx.callback?

File: python/mxnet/model.py (hunk 1 above)
Comment (piiswrong) [inline python/mxnet/model.py]: This looks like a callback function and should be put under python/callback.py

**Comment by leopd** [comment]
`bokeh=None` is cleaner.  I like that better than having to assert that it loaded every time.  It's not super friendly to end-users though.  They'll get an error message like `AttributeError: 'NoneType' object has no attribute 'figure'`.

Instead you could use:

```
try:
    import bokeh
except ImportError:
    class Bokeh_Failed_To_Import: pass
    bokeh = Bokeh_Failed_To_Import
```

then customers will see a message like `AttributeError: class Bokeh_Failed_To_Import has no attribute 'figure'` which is much easier to figure out how to fix.

**Comment by leopd** [comment]
I think Eric is saying that the class `LogValidationMetricsCallback` should be in the file https://github.com/dmlc/mxnet/blob/master/python/mxnet/callback.py

**Comment by kevinthesun** [comment]
OK. I'll fix these issues.

**Pull request closed without merging**